src/main/java/io/confluent/kafkarest/ConsumerState.java

```diff
@@ -15,114 +15,240 @@
+   * Gets the value deserializer for the Kafka consumer.
    */
-  protected abstract Decoder<KafkaV> getValueDecoder();
+  protected abstract Deserializer<KafkaV> getValueDeserializer();
 
   /**
-   * Converts a MessageAndMetadata using the Kafka decoder types into a ConsumerRecord using the
+   * Converts a ConsumerRecord using the Kafka deserializer types into a AbstractConsumerRecord using the
    * client's requested types. While doing so, computes the approximate size of the message in
    * bytes, which is used to track the approximate total payload size for consumer read responses to
    * determine when to trigger the response.
    */
-  public abstract ConsumerRecordAndSize<ClientK, ClientV> createConsumerRecord(
-      MessageAndMetadata<KafkaK, KafkaV> msg);
+  public abstract ConsumerRecordAndSize<ClientK, ClientV> convertConsumerRecord(
+      ConsumerRecord<KafkaK, KafkaV> msg);
 
   /**
    * Start a read on the given topic, enabling a read lock on this ConsumerState and a full lock on
-   * the ConsumerTopicState.
+   * the ConsumerSubscriptionState.
    */
-  public void startRead(ConsumerTopicState<KafkaK, KafkaV, ClientK, ClientV> topicState) {
-    lock.readLock().lock();
-    topicState.lock();
+  public void startRead() {
+    lock.lock();
+    heartbeatLock.lock();
   }
 
   /**
-   * Finish a read request, releasing the lock on the ConsumerTopicState and the read lock on this
+   * Finish a read request, releasing the lock on the ConsumerSubscriptionState and the read lock on this
    * ConsumerState.
    */
-  public void finishRead(ConsumerTopicState<KafkaK, KafkaV, ClientK, ClientV> topicState) {
-    topicState.unlock();
-    lock.readLock().unlock();
+  public void finishRead() {
+    // after finishing read change heartbeat time
+    this.nextHeartbeatTime = config.getTime().milliseconds() + heartbeatDelay;
+    lock.unlock();
+    heartbeatLock.unlock();
+  }
+
+  public boolean readStarted() {
+    return lock.isLocked();
+  }
+
+  public ConsumerReadTask clearFailedTask() {
+    ConsumerReadTask t = failedTask;
+    failedTask = null;
+    return t;
+  }
+
+  public void setFailedTask(ConsumerReadTask failedTask) {
+    this.failedTask = failedTask;
+  }
+
+  public Map<TopicPartition, OffsetAndMetadata> getConsumedOffsets() {
+    return consumedOffsets;
+  }
+
+  public Map<TopicPartition, OffsetAndMetadata> getCommittedOffsets() {
+    return committedOffsets;
   }
 
   public List<TopicPartitionOffset> commitOffsets() {
-    lock.writeLock().lock();
+    lock.lock();
     try {
-      consumer.commitOffsets();
       List<TopicPartitionOffset> result = getOffsets(true);
       return result;
     } finally {
-      lock.writeLock().unlock();
+      lock.unlock();
+    }
+  }
+
+  /**
+   * To keep a Consumer subscribed the poll must be invoked periodically
+   * to send heartbeats to the kafka back-end
+   * This method should not block for a long period.
+   */
+  public void sendHeartbeat() {
```

Comment:
It might be worth the wait for https://issues.apache.org/jira/browse/KAFKA-3888 to be done, as the heartbeat is being separated from the _poll_ method and will done in a separate background thread,  it would make this code much cleaner.